### PR TITLE
Move coprocessor to read pool

### PR DESCRIPTION
**Pull request by breeswish**
## What have you changed? (mandatory)

This PR moves coprocessor to read pool and removes the end-point-worker.

TODO:

- [x] Merge https://github.com/tikv/tikv/pull/3495 (test_coprocessor prelude)
- [x] Merge https://github.com/tikv/tikv/pull/3527 (async / await prelude)
- [x] Slow log
- [x] Support tracker in outdated situation
- [x] Add tests for trackers
- [x] ~~More outdated tests~~
- [x] Benchmark

## What are the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

CI

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

TODO

## Benchmark result if necessary (optional)

ycsb coprocessor QPS increased from 9W to 12W.

### CODE REVIEW

**Comment by solotzg** [comment]
LGTM

File: src/coprocessor/coprocessor.rs [hunk 1]
```diff
@@ -0,0 +1,1142 @@
...
+                            let yielded = Either::Left(resp);
+                            let next_state = Some((tracker, handler, finished));
+                            Some(Ok((yielded, next_state)))
+                        }
+                        None => {
+                            // Finished
+                            None
+                        }
+                    }
+                }).filter_map(|resp_or_tracker| match resp_or_tracker {
```
Comment (solotzg) [inline src/coprocessor/coprocessor.rs]: Is `Either` necessary here？Why not just handle the `tracker` directly in `stream::unfold`.

Comment (breeswish) [reply src/coprocessor/coprocessor.rs]: It is not necessary, only to be more readable.

**Comment by siddontang** [comment]
@solotzg 

Please approve it

File: src/coprocessor/coprocessor.rs [hunk 2]
```diff
@@ -0,0 +1,1142 @@
...
+use coprocessor::dag::executor::ExecutorMetrics;
+use coprocessor::metrics::*;
+use coprocessor::tracker::Tracker;
+use coprocessor::util as cop_util;
+use coprocessor::*;
+
+const OUTDATED_ERROR_MSG: &str = "request outdated.";
+const BUSY_ERROR_MSG: &str = "server is busy (coprocessor full).";
+
+pub struct Coprocessor<E: Engine> {
```
Comment (siddontang) [inline src/coprocessor/coprocessor.rs]: I remember that @BusyJay disagreed with the Coprocessor Name here. 

seem we should use a better name.

File: src/coprocessor/coprocessor.rs [hunk 3]
```diff
@@ -0,0 +1,1142 @@
...
+        peer: Option<String>,
+        is_streaming: bool,
+    ) -> (RequestHandlerBuilder<E::Snap>, ReqContext) {
+        match self.try_parse_request(req, peer, is_streaming) {
+            Ok(v) => v,
+            Err(err) => {
+                // If there are errors when parsing requests, create a dummy request handler.
+                let builder =
+                    box |_, _: &_| Ok(cop_util::ErrorRequestHandler::new(err).into_boxed());
+                let req_ctx = ReqContext::new(
```
Comment (siddontang) [inline src/coprocessor/coprocessor.rs]: oh, so we need to get the async snapshot for the invalid request and send this to execute in the read pool too?

File: src/coprocessor/coprocessor.rs [hunk 4]
```diff
@@ -0,0 +1,1142 @@
...
+        handler_builder: RequestHandlerBuilder<E::Snap>,
+    ) -> impl Future<Item = coppb::Response, Error = Error> {
+        // When this function is being executed, it may be queued for a long time, so that
+        // deadline may exceed.
+        future::result(tracker.req_ctx.deadline.check_if_exceeded())
+            .and_then(move |_| {
+                Self::async_snapshot(engine, &tracker.req_ctx.context)
+                    .map(|snapshot| (tracker, snapshot))
+            })
+            .and_then(move |(tracker, snapshot)| {
```
Comment (siddontang) [inline src/coprocessor/coprocessor.rs]: why repeat `.and_then(move |(tracker, snapshot)| {` 3 times?

File: src/coprocessor/coprocessor.rs [hunk 5]
```diff
@@ -0,0 +1,1142 @@
...
+        &self,
+        req: coppb::Request,
+        peer: Option<String>,
+    ) -> impl Stream<Item = coppb::Response, Error = ()> {
+        let (handler_builder, req_ctx) = self.parse_request(req, peer, true);
+        self.handle_stream_request(req_ctx, handler_builder)
+    }
+}
+
+fn make_tag(is_table_scan: bool) -> &'static str {
```
Comment (siddontang) [inline src/coprocessor/coprocessor.rs]: add inline

File: src/coprocessor/mod.rs [hunk 6]
```diff
@@ -53,59 +54,112 @@ trait RequestHandler: Send {
...
 }
 
-pub use self::dag::{ScanOn, Scanner};
-pub use self::endpoint::{
-    Host as EndPointHost, RequestTask, Task as EndPointTask, DEFAULT_REQUEST_MAX_HANDLE_SECS,
-};
+/// Denotes for a function that builds a `RequestHandler`.
+/// Due to rust-lang#23856, we have to make it a type alias of `Box<..>`.
+type RequestHandlerBuilder<Snap> =
+    Box<dyn for<'a> FnBox(Snap, &'a ReqContext) -> Result<Box<dyn RequestHandler + Send>> + Send>;
```
Comment (siddontang) [inline src/coprocessor/mod.rs]: em, why add `dyn for` here?

File: src/coprocessor/dag/expr/ctx.rs [hunk 7]
```diff
@@ -91,62 +91,68 @@ impl EvalConfig {
...
     }
 
-    pub fn set_ignore_truncate(mut self, new_value: bool) -> Self {
+    pub fn from_flags(flags: u64) -> Self {
+        let mut config = Self::new();
+        config.set_by_flags(flags);
+        config
+    }
+
+    pub fn set_ignore_truncate(&mut self, new_value: bool) -> &mut Self {
```
Comment (siddontang) [inline src/coprocessor/dag/expr/ctx.rs]: why change move to mutable ref?

File: src/coprocessor/tracker.rs [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+use util::time::{self, Duration, Instant};
+
+use coprocessor::dag::executor::ExecutorMetrics;
+use coprocessor::*;
+
+// If handle time is larger than the lower bound, the query is considered as slow query.
+const SLOW_QUERY_LOWER_BOUND: f64 = 1.0; // 1 second.
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum TrackerState {
```
Comment (siddontang) [inline src/coprocessor/tracker.rs]: can you split your PR? Your PR is too huge.

Maybe you can move tracker to another PR at first.

File: src/coprocessor/coprocessor.rs (hunk 3 above)
Comment (breeswish) [reply src/coprocessor/coprocessor.rs]: yes. However invalid requests should be rare, so the cost is affordable.

File: src/coprocessor/mod.rs (hunk 6 above)
Comment (breeswish) [reply src/coprocessor/mod.rs]: dyn is the new (recommended) syntax for specifying trait objects. If you don't write it, it will be implicitly written. This just make it clear. See https://github.com/rust-lang/rfcs/blob/master/text/2113-dyn-trait-syntax.md

Comment (breeswish) [reply src/coprocessor/mod.rs]: `for<'a>` is a HRTB here, see https://doc.rust-lang.org/nomicon/hrtb.html

File: src/coprocessor/dag/expr/ctx.rs (hunk 7 above)
Comment (breeswish) [reply src/coprocessor/dag/expr/ctx.rs]: Because I found it too ugly when used in DAG.

Previous:

```rust
        let mut eval_cfg = EvalConfig::new().set_by_flags(req.get_flags());
        // We respect time zone name first, then offset.
        if req.has_time_zone_name() && !req.get_time_zone_name().is_empty() {
            eval_cfg = box_try!(eval_cfg.set_time_zone_by_name(req.get_time_zone_name()));
        } else if req.has_time_zone_offset() {
            eval_cfg = box_try!(eval_cfg.set_time_zone_by_offset(req.get_time_zone_offset()));
        } else {
            // This should not be reachable. However we will not panic here in case
            // of compatibility issues.
        }
        if req.has_max_warning_count() {
            eval_cfg = eval_cfg.set_max_warning_cnt(req.get_max_warning_count() as usize);
        }
        if req.has_sql_mode() {
            eval_cfg = eval_cfg.set_sql_mode(req.get_sql_mode());
        }
        if req.has_is_strict_sql_mode() {
            eval_cfg = eval_cfg.set_strict_sql_mode(req.get_is_strict_sql_mode());
        }
```

Now:

```rust
        let mut eval_cfg = EvalConfig::from_flags(req.get_flags());
        // We respect time zone name first, then offset.
        if req.has_time_zone_name() && !req.get_time_zone_name().is_empty() {
            box_try!(eval_cfg.set_time_zone_by_name(req.get_time_zone_name()));
        } else if req.has_time_zone_offset() {
            box_try!(eval_cfg.set_time_zone_by_offset(req.get_time_zone_offset()));
        } else {
            // This should not be reachable. However we will not panic here in case
            // of compatibility issues.
        }
        if req.has_max_warning_count() {
            eval_cfg.set_max_warning_cnt(req.get_max_warning_count() as usize);
        }
        if req.has_sql_mode() {
            eval_cfg.set_sql_mode(req.get_sql_mode());
        }
        if req.has_is_strict_sql_mode() {
            eval_cfg.set_strict_sql_mode(req.get_is_strict_sql_mode());
        }
```

File: src/coprocessor/tracker.rs (hunk 8 above)
Comment (breeswish) [reply src/coprocessor/tracker.rs]: Ok

File: src/coprocessor/coprocessor.rs (hunk 4 above)
Comment (breeswish) [reply src/coprocessor/coprocessor.rs]: Because tracker is moved to closure inside, so it has to be passed out, otherwise it will be consumed, which is not what we want. Notice that tracker is constructed before future is executed, so it must be moved inside closure, otherwise when it is referenced it has been destroyed (which is incorrect and of course not allowed by Rust)

File: src/coprocessor/mod.rs (hunk 6 above)
Comment (Hoverbear) [reply src/coprocessor/mod.rs]: This is some seriously advanced lifetime usage. :+1:

Comment (breeswish) [reply src/coprocessor/mod.rs]: @Hoverbear Yes! And very useful here, otherwise you cannot write its type. (and there are compiler bugs 🤣

File: src/coprocessor/coprocessor.rs (hunk 4 above)
Comment (solotzg) [reply src/coprocessor/coprocessor.rs]: @breeswish 
How about making tracker and snapshot boxed? The size of `Future::Item` may be big currently.

Comment (solotzg) [reply src/coprocessor/coprocessor.rs]: @breeswish some tests have been done, and u can see them in TIKV-1489.

**Comment by huachaohuang** [comment]
@breeswish Seems the PR contains some unrelated changes like `EvalConfig`, can you separate that to another PR first?

File: src/coprocessor/endpoint.rs [hunk 9]
```diff
@@ -11,641 +11,448 @@
...
+        // When this function is being executed, it may be queued for a long time, so that
+        // deadline may exceed.
+
+        let tracker_and_handler_future = future::result(
+            tracker.req_ctx.deadline.check_if_exceeded(),
+        ).and_then(move |_| {
+            Self::async_snapshot(engine, &tracker.req_ctx.context)
+                .map(|snapshot| (tracker, snapshot))
+        })
+            .and_then(move |(tracker, snapshot)| {
```
Comment (hicqu) [inline src/coprocessor/endpoint.rs]: Is the odd format really processed by rustfmt?

File: src/coprocessor/endpoint.rs [hunk 10]
```diff
@@ -11,641 +11,448 @@
...
+                // When snapshot is retrieved, deadline may exceed.
+                future::result(tracker.req_ctx.deadline.check_if_exceeded())
+                    .map(|_| (tracker, snapshot))
+            })
+            .and_then(move |(tracker, snapshot)| {
+                future::result(handler_builder.call_box((snapshot, &tracker.req_ctx)))
+                    .map(|handler| (tracker, handler))
+            });
+
+        stream::futures_ordered(vec![tracker_and_handler_future])
```
Comment (hicqu) [inline src/coprocessor/endpoint.rs]: Why not `tracker_and_handler_future.map(...).flatten_stream`?

File: src/coprocessor/endpoint.rs [hunk 11]
```diff
@@ -11,641 +11,448 @@
...
-                req.tracker.task_count(task_count);
+    /// Handle a stream request and run on the read pool. Returns a stream producing each
+    /// result, which must be a `Response` and will never fail. If there are errors during
+    /// handling, they will be embedded in the `Response`.
+    fn handle_stream_request(
+        &self,
+        req_ctx: ReqContext,
+        handler_builder: RequestHandlerBuilder<E::Snap>,
+    ) -> impl Stream<Item = coppb::Response, Error = ()> {
+        let (tx, rx) = mpsc::channel::<coppb::Response>(self.stream_channel_size);
```
Comment (hicqu) [inline src/coprocessor/endpoint.rs]: How about use `util::mpsc::channel`? According to jay's bench, it's much better than futures::Sender.

**Comment by hicqu** [comment]
rest LGTM.

File: src/coprocessor/endpoint.rs (hunk 9 above)
Comment (breeswish) [reply src/coprocessor/endpoint.rs]: yes :(

File: src/coprocessor/endpoint.rs (hunk 11 above)
Comment (breeswish) [reply src/coprocessor/endpoint.rs]: Ok, I will have a test

File: src/coprocessor/endpoint.rs (hunk 10 above)
Comment (breeswish) [reply src/coprocessor/endpoint.rs]: might not be applicable. tracker_and_handler_future is a future (of pair `(tracker, handler)`), not a future of a stream

**Comment by hicqu** [comment]
LGTM.
